src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -233,11 +233,21 @@ private static unsafe int GetResponseHeaderCharBufferLength(SafeWinHttpHandle re
             {
                 int lastError = Marshal.GetLastWin32Error();
 
-                Debug.Assert(lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND);
+                if (!isTrailingHeaders)
+                {
+                    Debug.Assert(lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND);
 
-                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+                    if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+                    {
+                        throw WinHttpException.CreateExceptionUsingError(lastError, nameof(Interop.WinHttp.WinHttpQueryHeaders));
+                    }
+                }
+                else
                 {
-                    throw WinHttpException.CreateExceptionUsingError(lastError, nameof(Interop.WinHttp.WinHttpQueryHeaders));
+                    if (!(lastError == Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER || lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND))
```

Comment:
Is `ERROR_WINHTTP_HEADER_NOT_FOUND` here to silently fail when `WINHTTP_QUERY_FLAG_TRAILERS` is unsupported?

Reply:
No, it is to handle HTTP responses that don't have any trailers.

A response will always have headers, but trailers are optional.

Reply:
In this library these checks are unprecedental, setting unsupported options usually leads to `WinHttpException`. I've seen checks on `Environment.OSVersion` in other BCL libs though .

Reply:
What if WinHttp decides to backport the feature to earlier OS versions? I would feel more comfortable with established pattern of relying on `WinHttpException`. Is that possible?

Reply:
With the current design, this feature is implicit (filling the collection whenever possible), so throwing is not an option.

However I think that checking the error code after the first `WinHttpQueryHeaders` shall be a sufficient (and actually better) solution.

Reply:
One thing to consider here --- we should be very careful about having an API that, depending on OS version:

a) works as expected.
b) appears to work, but has an empty collection with zero notice to user.

This will be very fragile.

Reply:
@scalablecory I tried to address this in my proposal in https://github.com/dotnet/runtime/issues/44778#issuecomment-756304550. I suggest to discuss the API under the issue, comments and ideas are more than welcome there.

Reply:
I discussed OS support detection for trailers with WinHttp team and this is their suggestion:

> How about this for detection.
> 
> You try to query WINHTTP_OPTION_STREAM_ERROR_CODE from you session handle.
> 
> If it fails with ERROR_INVALID_PARAMETER, there is no support for trailers.
> 
> Otherwise, it should fail ERROR_WINHTTP_INCORRECT_HANDLE_TYPE, as the option cannot be queried from session handles.

Reply:
> You try to query WINHTTP_OPTION_STREAM_ERROR_CODE from you session handle.

I guess these are distinct features which have been introduced in the same release right? What if they get misaligned because trailing headers get backported to an earlier version?

Reply:
I presume so. This is the WinHttp team's suggestion.